### DISCUSSION TITLE
Change from using ComparisonTemplateLibrary/ComparisonLibrary to ComparisonLevels Results in Issue w/ estimating u values

### DISCUSSION THREAD

**Discussion by beedrumms** [moj-analytical-services/splink#1896] [Q&A] [answered]
Apologies for the long winded issue -- I have been using the ComparisonTemplateLibrary/ComparisonLibrary for specifying my Comparisons but **switched to ComparisonLevels** to incorporate some complex business logic. This switch has resulted in **u value estimation issues** when i run linker.estimate_u_using_random_sampling(). The issue is that, with the switch, suddenly comparison levels are "never observed in the training data". However, the same comparisons work w/ Template. 

**I expect that there is something wrong w/ my sql conditions** (shown below) **but i have tried so many different iterations to make it right** but am getting no where. Please lmk if you have any thoughts / suggestions to fix this issue! :)  

Env: 
- Azure Databricks -- RT 12.2 -- 56gb ram -- 8 workers
- Python 3.10 

Recreate Issue: 
`first_name_comparison = {
    'output_column_name': 'First_Name',
    'comparison_levels': [
        {
            'sql_condition': '"First_Name_l" IS NULL OR "First_Name_r" IS NULL',
            'label_for_charts': 'Null',
            'is_null_level': True
        },
        {
            'sql_condition': '"First_Name_l" = "First_Name_r"',
            'label_for_charts': 'Exact match',
            'tf_adjustment_column': 'First_Name',
            'tf_adjustment_weight': 1.0
        },
        {
            'sql_condition': 'damerau_levenshtein("First_Name_l", "First_Name_r") <= 1',
            'label_for_charts': 'Damerau_levenshtein <= 1'
        },
        {
            'sql_condition': 'jaro_winkler("First_Name_l", "First_Name_r") >= 0.9',
            'label_for_charts': 'Jaro_winkler_similarity >= 0.9'
        },
        {
            'sql_condition': 'jaro_winkler("First_Name_l", "First_Name_r") >= 0.8',
            'label_for_charts': 'Jaro_winkler >= 0.8'
        },
        {
            'sql_condition': 'ELSE', 
            'label_for_charts': 'All other comparisons'
        }],
    'comparison_description': 'Exact match vs. anything else'
}`

`last_name_comparison = {
    'output_column_name': 'Last_Name',
    'comparison_levels': [
        {
            'sql_condition': '"Last_Name_l" IS NULL OR "Last_Name_r" IS NULL',
            'label_for_charts': 'Null',
            'is_null_level': True
        },
        {
            'sql_condition': '"Last_Name_l" = "Last_Name_r"',
            'label_for_charts': 'Exact match'
        },
        {
            'sql_condition': 'damerau_levenshtein("Last_Name_l", "Last_Name_r") <= 1',
            'label_for_charts': 'Damerau_levenshtein <= 1'
        },
        {
            'sql_condition': 'jaro_winkler("Last_Name_l", "Last_Name_r") >= 0.9',
            'label_for_charts': 'Jaro_winkler >= 0.9'
        },
        {
            'sql_condition': 'jaro_winkler("Last_Name_l", "Last_Name_r") >= 0.8',
            'label_for_charts': 'Jaro_winkler >= 0.8'
        },
        {
            'sql_condition': 'ELSE', 'label_for_charts': 'All other comparisons'
        }],
         'comparison_description': 'Exact match vs. Last_Name within levenshtein threshold 1 vs. Last_Name within damerau-levenshtein threshold 1 vs. Last_Name within jaro_winkler thresholds 0.9, 0.8 vs. anything else'
}`

`dob_comparison = {
    'output_column_name': 'dob',
    'comparison_levels': [
        {
            'sql_condition': '"dob_l" IS NULL OR "dob_r" IS NULL',
            'label_for_charts': 'Null',
            'is_null_level': True
        },
        {
            'sql_condition': '"dob_l" = "dob_r"',
            'label_for_charts': 'Exact match'
        },
        {
            'sql_condition': "regexp_extract(`dob_l`, '^[0-9]{1,4}', 0) = regexp_extract(`dob_r`, '^[0-9]{1,4}', 0)", 
            'label_for_charts': 'Exact match on Year'
        },
        {
            'sql_condition': "(`dob_l` = `dob_r`) AND (NOT (SUBSTR(dob_l, -5) = '01-01'))",
            'label_for_charts': 'Exact match and not the 1st Jan'
        },
        {
            'sql_condition': "(`dob_l` = `dob_r`) AND (NOT (SUBSTR(dob_l, -5) = '07-01'))", 
            'label_for_charts': 'Exact match and not the 1st July'
        },
        {
            'sql_condition': 'ELSE', 'label_for_charts': 'All other comparisons'
        }],
    'comparison_description': 'Exact match vs. dob within levenshtein thresholds 1, vs. anything else'
}`

`health_number_comparison = {
    'output_column_name': 'Health_Number',
    'comparison_levels': [{'sql_condition': '"Health_Number_l" IS NULL OR "Health_Number_r" IS NULL',
    'label_for_charts': 'Null',
    'is_null_level': True},
    {
        'sql_condition': '"Health_Number_l" = "Health_Number_r"',
        'label_for_charts': 'Exact match'
    },
    {
        'sql_condition': 'levenshtein("Health_Number_l", "Health_Number_r") <= 1',
        'label_for_charts': 'Levenshtein <= 1'
    },
    {
        'sql_condition': "(jaro_winkler(`First_Name_l`, `First_Name_r`) >= 0.9) AND (jaro_winkler(`Last_Name_l`, `Last_Name_r`) >= 0.9) AND (regexp_extract(`dob_l`, '^[0-9]{1,4}', 0) = regexp_extract(`dob_r`, '^[0-9]{1,4}', 0)) AND (NOT (SUBSTR(Health_Number_l, 1) = '0'))", 
        'label_for_charts': 'L Red Card Exception'
    },
    {
        'sql_condition': "(jaro_winkler(`First_Name_l`, `First_Name_r`) >= 0.9) AND (jaro_winkler(`Last_Name_l`, `Last_Name_r`) >= 0.9) AND (regexp_extract(`dob_l`, '^[0-9]{1,4}', 0) = regexp_extract(`dob_r`, '^[0-9]{1,4}', 0)) AND (NOT (SUBSTR(Health_Number_r, 1) = '0'))", 
        'label_for_charts': 'R Red Card Exception'
    },
    {
        'sql_condition': 'ELSE', 
        'label_for_charts': 'All other comparisons'
    }],
    'comparison_description': 'Exact match vs. levenshtein threshold 1, Null vs. Red Card exception vs. anything else'
}`

`settings = {

    "link_type": "link_and_dedupe",

    "blocking_rules_to_generate_predictions": [
        block_on(["Health_Number"]),
        block_on(["First_Name", "dob"]),
        block_on(["Last_Name", "dob"]),
        block_on(["First_Name", "Last_Name"]),
    ],

    "comparisons": [
        first_name_comparison, 
        last_name_comparison, 
        dob_comparison,  
        health_number_comparison,
    ],

    "retain_matching_columns": True, 
    
    "retain_intermediate_calculation_columns": True,

    "em_convergence": 0.01,

    "max_iterations": 25,

    "unique_id_column_name": 'unique_id',

    "sql_dialect": 'spark'
}`

`spark.sql("USE CATALOG XXX;")`
`spark.sql("USE SCHEMA XXX;")`

`
linker = SparkLinker([XX, X], 
                     settings_dict=settings,
                     input_table_aliases=["XX", "X"],
                     validate_settings=True,
                     spark=spark, 
                     catalog="XXX",
                     )
`

`
deterministic_rules = ["l.Health_Number = r.Health_Number",
                       "l.First_Name = r.First_Name and l.Last_Name = r.Last_Name and l.dob = r.dob",
                       ]
    linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.55)
`

`for i in range(3):
    
    with warnings.catch_warnings():
        
        warnings.filterwarnings("ignore", category=DeprecationWarning)

        linker.estimate_u_using_random_sampling(max_pairs=1e8)` ### THIS IS WHERE THE ISSUE OCCURS
`

EXAMPLE OUTPUT: 
----- Estimating u probabilities using random sampling -----
u probability not trained for First_Name - Exact match (comparison vector value: 4). This usually means the comparison level was never observed in the training data.
u probability not trained for First_Name - Jaro_winkler_similarity >= 0.9 (comparison vector value: 2). This usually means the comparison level was never observed in the training data.
u probability not trained for First_Name - Jaro_winkler >= 0.8 (comparison vector value: 1). This usually means the comparison level was never observed in the training data.
u probability not trained for First_Name - All other comparisons (comparison vector value: 0). This usually means the comparison level was never observed in the training data.
u probability not trained for Last_Name - Exact match (comparison vector value: 4). This usually means the comparison level was never observed in the training data.
u probability not trained for Last_Name - Jaro_winkler >= 0.9 (comparison vector value: 2). This usually means the comparison level was never observed in the training data.
u probability not trained for Last_Name - Jaro_winkler >= 0.8 (comparison vector value: 1). This usually means the comparison level was never observed in the training data.
u probability not trained for Last_Name - All other comparisons (comparison vector value: 0). This usually means the comparison level was never observed in the training data.
u probability not trained for dob - Exact match (comparison vector value: 4). This usually means the comparison level was never observed in the training data.
u probability not trained for dob - Exact match and not the 1st Jan (comparison vector value: 2). This usually means the comparison level was never observed in the training data.
u probability not trained for dob - Exact match and not the 1st July (comparison vector value: 1). This usually means the comparison level was never observed in the training data.
u probability not trained for Health_Number - Exact match (comparison vector value: 4). This usually means the comparison level was never observed in the training data.
u probability not trained for Health_Number - L Red Card Exception (comparison vector value: 2). This usually means the comparison level was never observed in the training data.
u probability not trained for Health_Number - R Red Card Exception (comparison vector value: 1). This usually means the comparison level was never observed in the training data.
u probability not trained for Health_Number - All other comparisons (comparison vector value: 0). This usually means the comparison level was never observed in the training data.

**Comment by ADBond** [comment, marked as answer]
I think the issue is that you are using double quotes for column names - in SparkSQL these [are interpreted as string literals](https://spark.apache.org/docs/latest/sql-ref-literals.html#string-literal). So for instance in your `last_name_comparison`, from the logs it seems the only comparison level that has _any_ observed values is `Damerau_levenshtein <= 1` (as all the others are mentioned in the warnings). This is because it is _always_ true that `damerau_levenshtein("Last_Name_l", "Last_Name_r") <= 1`, as these strings only differ by a character (Spark seeing e.g. `"Last_Name_l"` as a string literal, and at no point associating this with a table in the database), so every record pair gets put into this category regardless of what the actual data looks like (and so nothing ever gets put into any of the other levels). 

To fix this you should use backticks to refer to columns in Spark, so for instance:
```py
{ 'sql_condition': 'damerau_levenshtein(`Last_Name_l`, `Last_Name_r`) <= 1', 'label_for_charts': 'Damerau_levenshtein <= 1' }
```

and similarly for the other levels + comparisons. This is one of the things that Splink does automatically when using the libraries, but is easy to get wrong when writing custom levels.

I haven't checked if this is the problem for all comparisons, so let me know if there are still training issues after doing this and I can see if there is anything else going wrong

**Reply by beedrumms** [reply]
@ADBond thank you so much! terrible oversight on my part lol -- I updated and the **name comparisons are working perfectly**. Still having some trouble w/ my custom comparisons but i expect that they are not configured properly -- I will work through it and repost the working sql code. Thanks again!

**Reply by ADBond** [reply]
No trouble, easily done 👍
